[pt][quant] Support either min or max in qclamp

dskhudia opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45937 [pt][quant] Support either min or max in qclamp**

torch.clamp can now be used with quantized tensors with either min argument or max argument only

Fixes https://github.com/pytorch/pytorch/issues/45928

Differential Revision: [D24153431](https://our.internmc.facebook.com/intern/diff/D24153431/)

dskhudia committed:
[pt][quant] Support either min or max in qclamp

torch.clamp can now be used with quantized tensors with either min argument or max argument only

Fixes https://github.com/pytorch/pytorch/issues/45928

Differential Revision: [D24153431](https://our.internmc.facebook.com/intern/diff/D24153431/)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qclamp.cpp:
```diff
@@ -84,14 +86,18 @@ Tensor quantized_clamp_impl(
   if (min && max) {
 #ifdef USE_PYTORCH_QNNPACK
-    if (at::globalContext().qEngine() == at::QEngine::QNNPACK && qx.scalar_type() == kQUInt8) {
+    if (at::globalContext().qEngine() == at::QEngine::QNNPACK &&
```
qnnpack does not support the other two cases, can you keep the original error for qnnpack?

dskhudia replied:
Does the following look ok?

jerryzh168 replied:
maybe you can separate the checks for qnnpack and fbgemm to make it more clear..

jerryzh168 replied:
i.e. don't repeat USE_PYTORCH_QNNPACK

jerryzh168 reviewed (approved):
Thanks! please keep the original error for qnnpack.

dskhudia committed:
Update on "[pt][quant] Support either min or max in qclamp"


torch.clamp can now be used with quantized tensors with either min argument or max argument only

Fixes https://github.com/pytorch/pytorch/issues/45928

Differential Revision: [D24153431](https://our.internmc.facebook.com/intern/diff/D24153431/)

[ghstack-poisoned]

dskhudia committed:
Update on "[pt][quant] Support either min or max in qclamp"


torch.clamp can now be used with quantized tensors with either min argument or max argument only

Fixes https://github.com/pytorch/pytorch/issues/45928

Differential Revision: [D24153431](https://our.internmc.facebook.com/intern/diff/D24153431/)

[ghstack-poisoned]

dskhudia committed:
Update on "[pt][quant] Support either min or max in qclamp"


torch.clamp can now be used with quantized tensors with either min argument or max argument only

Fixes https://github.com/pytorch/pytorch/issues/45928

Differential Revision: [D24153431](https://our.internmc.facebook.com/intern/diff/D24153431/)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qclamp.cpp:
```diff
@@ -84,14 +86,23 @@ Tensor quantized_clamp_impl(
   } else {
-    TORCH_CHECK(
-        false, "Both min and max should be specified for quantized clamp!");
+    if (at::globalContext().qEngine() == at::QEngine::QNNPACK) {
```
maybe add a USE_PYTORCH_QNNPACK macro here
```
#ifdef USE_PYTORCH_QNNPACK
    if (at::globalContext().qEngine() == at::QEngine::QNNPACK) {
     ...
   }
#endif
// fallback to fbgemm
if (max) {
} else if (min) {
} else {
}
```

dskhudia committed:
Update on "[pt][quant] Support either min or max in qclamp"


torch.clamp can now be used with quantized tensors with either min argument or max argument only

Fixes https://github.com/pytorch/pytorch/issues/45928

Differential Revision: [D24153431](https://our.internmc.facebook.com/intern/diff/D24153431/)

[ghstack-poisoned]

dskhudia committed:
Update on "[pt][quant] Support either min or max in qclamp"


torch.clamp can now be used with quantized tensors with either min argument or max argument only

Fixes https://github.com/pytorch/pytorch/issues/45928

Differential Revision: [D24153431](https://our.internmc.facebook.com/intern/diff/D24153431/)

[ghstack-poisoned]

dskhudia committed:
Update on "[pt][quant] Support either min or max in qclamp"


torch.clamp can now be used with quantized tensors with either min argument or max argument only

Fixes https://github.com/pytorch/pytorch/issues/45928

Differential Revision: [D24153431](https://our.internmc.facebook.com/intern/diff/D24153431/)

[ghstack-poisoned]
